Use templates for common geometric constants

rubiefawn opened the pull request:
`lmms_constants.h` has several common constants such as `PI` and `E`, with variants of each for `float`, `double`, and `long double`. Many of these are currently unused in the repo but are of potential use. This change replaces `LD_PI`, `D_PI`, and `F_PI` with `lmms::numbers::pi`, and so on, in the style of C++20's `<numbers>`.

Currently, the constants are set using literal values, but once C++20 is fully ready these can be changed to their `std::numbers::*` equivalents instead.

messmerd commented:
You should use `inline constexpr` since the variables are in a header, though maybe it would be best to wait for C++20 (#7554) and GCC 10 where we'll get the [`<numbers>` header](https://en.cppreference.com/w/cpp/header/numbers).

rubiefawn commented:
Updated to account for `<numbers>` and C++20:
- Marked all `lmms_constants.h` constants with an exact equivalent in `<numbers>` as deprecated
- Removed all `lmms_constants.h` constants where no variant is currently in use
- Using `inline constexpr`
- Using `std::floating_point` concept instead of `typename`

messmerd commented:
This won't work. I suggested *waiting* for C++20 and the compiler upgrade since we are in the process of doing that now. 

Currently we are using C++17 and GCC 9.3 for Linux and MinGW builds, so we can't use the <numbers> header or concepts yet.

If you wanted, maybe you could convert our mathematical constants to use the same names as those in <numbers>, but in the `lmms` namespace. Then once we upgrade to C++20, switching to <numbers> will be as easy as swapping out `lmms::pi_v<double>` for `std::numbers::pi_v<double>`. That may be quite a bit of work for you though, so it's up to you.

rubiefawn commented:
I'm aware that we don't have C++20 ready yet. I made the suggested changes (and more) in advance under the assumption that this PR was now dependent on #7554, based off of your suggestion. If this is not the case I can change it back, I'm good with whatever.

My original intention with this PR was small in scope; I just wanted to clean up `lmms_constants.h`. That being said, I love your suggestion for `lmms::numbers`, and if updating all usage of the constants is on the table, I'd love to explore that.

messmerd commented:
Ah gotcha. Sorry for the miscommunication. The C++20 support could happen tomorrow if people could review the PR, but upgrading our MinGW compiler and dependencies will be tricky and take some time.

I think the `lmms::numbers` idea would be good to implement. Function scope usages could use `using namespace numbers;` so that it doesn't need to be typed out fully as the clunkier `numbers::pi` or `numbers::pi_v<float>` in places where that would look better.
I'm not sure how big a task it would be, but it would probably be fairly easy with vscode or some other program to just search through `src/`, `include/`, and `plugins/` and replace all.

sakertooth commented:
If I understand @messmerd correctly, I agree with him. We should wait for C++20, remove `lmms_constants.h`, and then just use the numeric library provided by the standard. The more code we can cut down and remove, I see as a win.

rubiefawn commented:
Just to clarify, there are several useful constants in `lmms_constants.h` that have no overlap with C++20 `<numbers>`. This PR isn't about removing that header entirely, it's just to clean it up. I agree with the "less is more" sentiment though.

On that note, I noticed `panning_constants.h` and `panning.h` and am wondering if those can be merged into one file (or even just moved to `lmms_constants.h`). Those could be considered further candidates for constant cleanup.

The current plan is to wait for #7554 before I make further changes. Input is welcome in the meantime!

sakertooth commented:
#7554 is here, so you can try and make the switch to the numbers header @rubiefawn (we have partial C++20 support, but I think we should have the numbers header at least).

> On that note, I noticed panning_constants.h and panning.h and am wondering if those can be merged into one file (or even just moved to lmms_constants.h). Those could be considered further candidates for constant cleanup.

It's up to you really (or anyone else that wants to give it a go). I personally think it could go multiple ways with no apparent downside.

messmerd commented:
@sakertooth Nope, like most C++20 features, we're waiting for GCC 10.

See: https://en.cppreference.com/w/cpp/compiler_support/20
(Mathematical constants P0631R8)

sakertooth commented:
Hard to classify it as an upgrade then, but I guess it's a step in the right direction..

messmerd commented:
I'm using both class type NTTP and designated initializers in the pin connector now, so it's not nothing. Once I'm done with the pin connector, I'd like to upgrade to GCC 11 on Linux and MinGW 13 for the cross-compiled Windows builds. Then we'll have nearly all C++20 features including `<numbers>`.

rubiefawn commented:
This is the second time I've gotten duplicated commit history after pushing. Not sure if GitHub just doesn't like rebase or if it's a skill issue (likely the latter). How do I avoid this in the future?

Rossmaxx commented:
Don't worry, we do squash on merge.

Rossmaxx commented:
Wait, seems like lots of unnecessary diffs (from previously merged prs)

Rossmaxx commented on include/BasicFilters.h:
```diff
@@ -774,7 +773,7 @@ class BasicFilters
 			m_p = ( 3.6f - 3.2f * f ) * f;
 			m_k = 2.0f * m_p - 1;
-			m_r = _q * powf( F_E, ( 1 - m_p ) * 1.386249f );
+			m_r = _q * powf(numbers::e_v<float>, (1 - m_p) * 1.386249f);
```
```suggestion
			m_r = _q * std::exp((1 - m_p) * 1.386249f);
```
One simplification.

Rossmaxx commented on include/BasicFilters.h:
```diff
@@ -791,7 +790,7 @@ class BasicFilters
 			m_p = ( 3.6f - 3.2f * f ) * f;
 			m_k = 2.0f * m_p - 1.0f;
-			m_r = _q * 0.1f * powf( F_E, ( 1 - m_p ) * 1.386249f );
+			m_r = _q * 0.1f * powf(numbers::e_v<float>, (1 - m_p) * 1.386249f);
```
```suggestion
			m_r = _q * 0.1f * std::exp((1 - m_p) * 1.386249f);
```
Here too.

Rossmaxx commented on include/lmms_constants.h:
```diff
@@ -25,36 +25,54 @@
 #ifndef LMMS_CONSTANTS_H
 #define LMMS_CONSTANTS_H
 
-namespace lmms
```
FYI, The C++ 20 PR was merged, so might as well make the switch now.

rubiefawn replied:
Not entirely: https://github.com/LMMS/lmms/pull/7558#issuecomment-2495622443. Concepts and mathematical constants are waiting on GCC 10+. Looks like that is planned for after #7459 (see https://github.com/LMMS/lmms/pull/7558#issuecomment-2495678199), but I made these changes with the intention that they won't be dependent on the compiler upgrade.

Rossmaxx replied:
Ok then.

Rossmaxx commented on plugins/Eq/EqCurve.cpp:
```diff
@@ -527,10 +522,8 @@ void EqHandle::setlp48()
-
-	const double w = 2 * LD_PI * freq / SR ;
-	const double PHI = pow( sin( w / 2 ), 2 ) * 4;
+	const double w = sin(numbers::pi * freq / Engine::audioEngine()->outputSampleRate());
```
You sure about this?

Rossmaxx replied:
There seems to be a change in mathematical calculation. That's what I meant.

Rossmaxx replied:
Nvm did some pen paper testing and the math checks out.

Rossmaxx replied:
Also there's one thing you can do, make `outputSampleRate` into a member variable with initialisation in the constructor to avoid calling the function every time. It gets a bit lengthy.

rubiefawn replied:
I'd prefer to avoid making changes beyond the immediate scope of lines that already have to be updated to accommodate `lmms::numbers::*`, since that's what this PR is focused on. Is that okay?

Rossmaxx replied:
Ok then.

Rossmaxx reviewed (commented):
Some thoughts which came across when I looked at the diff.

Rossmaxx reviewed (approved):
Looks good. 
Untested but style and diff looks ok.

rubiefawn commented:
What else does this need before it can be merged?

Rossmaxx commented:
Some testing to check for regressions?

rubiefawn commented:
I've personally verified that all of the demo tracks play without any obvious defects, and nothing has exploded yet, but that's just me on my machine. Would you please add the 'needs testing' label to this?

How thorough and documented should testing this be beyond the [default tests](https://github.com/LMMS/lmms/wiki/Testing-Pull-Requests)?

Rossmaxx commented:
It should be ok to merge, but i'm just concerned about something breaking somewhere. I'll add the label tho.

rubiefawn commented:
In that case, I'll just also perform the same tests on my Windows machine and get someone in the Discord with a Mac to do the same. Would that be sufficient?

Rossmaxx commented:
@sakertooth @messmerd anyone mind dropping a second approval?

messmerd commented on include/lmms_constants.h:
A few constants here are not part of `std::numbers`. They are `tau`, `pi_half`, `pi_sqr`, and `inv_e`. Their usage would prevent us from upgrading to `std::numbers` once it is available.

Additionally, throughout this PR, the constants are used like this:
```c++
void foo()
{
	const auto var = numbers::pi_v<float> * numbers::pi_v<float> * bar;
}
```

This is alright, but upgrading to `std::numbers` would require prepending `std::` to every constant, making them longer and clunkier. The more common way to do it (especially when a lot of constants are used) would be this:
```c++
void foo()
{
	using namespace numbers; // local scope only
	const auto var = pi_v<float> * pi_v<float> * bar;
}
```

This can be more concise and easy to upgrade since we'd only need to swap out `#include "lmms_constants.h"` for `#include <numbers>` and `using namespace numbers;` for `using namespace std::numbers;` then delete `lmms_constants.h`.

Rossmaxx replied:
We can't delete the constants header even with the change. There are a few constants there which are beyond the scope of std::numbers.

messmerd replied:
Oh you're right, I didn't see the rest of the file. We'll at least be able to avoid including `lmms_constants.h` in cases that just require math calculations.

Rossmaxx replied:
I do plan to consolidate the rest of the file with `lmms_math.h` once this PR and full C++ 20 support goes through. Out of scope but wanted to mention.

messmerd reviewed (commented):
Sorry this review took so long

rubiefawn commented:
I just batch applied the suggestions related to std math functions, but it looks like something isn't playing nice with those. I'll take a closer look at it in a few hours.

rubiefawn commented:
It appears to be working now. There's still plenty of instances of C-style math functions, but at least the ones on the lines I already touched are updated to their C++ std counterparts. Those should probably be updated at some point (in fact, I'd be happy to do that as well) but in a different PR since it's beyond the scope of updating `lmms_constants.h`.

messmerd commented:
Thanks!
